File: packages/python/src/alumnium/mcp/drivers.py
"""Driver factory functions for different platforms."""

from os import getenv
from pathlib import Path
from typing import Any


def create_chrome_driver(capabilities: dict[str, Any], server_url: str | None, screenshot_dir: Path) -> Any:
    if getenv("ALUMNIUM_DRIVER", "selenium").lower() == "playwright":
        return create_playwright_driver(capabilities, screenshot_dir)
    else:
        return create_selenium_driver(capabilities, server_url)


def create_playwright_driver(capabilities: dict[str, Any], screenshot_dir: Path) -> Any:
    """Create async Playwright driver from capabilities."""
    import asyncio
    from threading import Thread

    from playwright.async_api import async_playwright

    # Create event loop in dedicated thread (shared by Playwright and driver)
    loop = asyncio.new_event_loop()
    thread = Thread(target=lambda: asyncio.set_event_loop(loop) or loop.run_forever(), daemon=True)
    thread.start()

    # Create Playwright resources in the shared event loop
    async def _create_resources():
        playwright = await async_playwright().start()
        headless = getenv("ALUMNIUM_PLAYWRIGHT_HEADLESS", "true").lower() == "true"
        browser = await playwright.chromium.launch(headless=headless)
        context = await browser.new_context(
            record_video_dir=screenshot_dir / "videos",
            extra_http_headers=capabilities.get("headers", {}),
        )

        await context.tracing.start(screenshots=True, snapshots=True, sources=True)
        page = await context.new_page()

        return page

    # Run resource creation in the shared loop
    future = asyncio.run_coroutine_threadsafe(_create_resources(), loop)
    page = future.result()
    return (page, loop)


def create_selenium_driver(capabilities: dict[str, Any], server_url: str | None) -> Any:
    """Create Selenium Chrome driver from capabilities."""
    from selenium.webdriver.chrome.options import Options

    headers = capabilities.pop("headers", {})
    options = Options()

    # Apply all capabilities to options
    for key, value in capabilities.items():
        if key != "platformName":
            options.set_capability(key, value)

    # Use Remote driver if server_url provided, otherwise local Chrome
    if server_url:
        from selenium.webdriver import Remote

        driver = Remote(command_executor=server_url, options=options)
    else:
        from selenium.webdriver import Chrome

        driver = Chrome(options=options)

    if headers:
        driver.execute_cdp_cmd("Network.enable", {})  # type: ignore[reportAttributeAccessIssue]
        driver.execute_cdp_cmd("Network.setExtraHTTPHeaders", {"headers": headers})  # type: ignore[reportAttributeAccessIssue]

    return driver


def create_ios_driver(capabilities: dict[str, Any], server_url: str | None) -> Any:
    """Create Appium iOS driver from capabilities."""
    from appium.options.ios import XCUITestOptions
    from appium.webdriver.client_config import AppiumClientConfig
    from appium.webdriver.webdriver import WebDriver as Appium

    options = XCUITestOptions()

    # Load capabilities into options
    options.load_capabilities(capabilities)

    # Determine server URL: parameter > env var > default
    if server_url:
        remote_server = server_url
    else:
        remote_server = getenv("ALUMNIUM_APPIUM_SERVER", "http://localhost:4723")

    # Set up Appium client config
    client_config = AppiumClientConfig(
        username=getenv("LT_USERNAME"),
        password=getenv("LT_ACCESS_KEY"),
        remote_server_addr=remote_server,
        direct_connection=True,
    )

    # Create Appium driver
    driver = Appium(client_config=client_config, options=options)

    return driver


def create_android_driver(capabilities: dict[str, Any], server_url: str | None) -> Any:
    """Create Appium Android driver from capabilities."""
    from appium.options.android import UiAutomator2Options
    from appium.webdriver.client_config import AppiumClientConfig
    from appium.webdriver.webdriver import WebDriver as Appium

    options = UiAutomator2Options()

    # Load capabilities into options
    options.load_capabilities(capabilities)

    # Determine server URL: parameter > env var > default
    if server_url:
        remote_server = server_url
    else:
        remote_server = getenv("ALUMNIUM_APPIUM_SERVER", "http://localhost:4723")

    # Set up Appium client config
    client_config = AppiumClientConfig(
        username=getenv("LT_USERNAME"),
        password=getenv("LT_ACCESS_KEY"),
        remote_server_addr=remote_server,
        direct_connection=True,
    )

    # Create Appium driver
    return Appium(client_config=client_config, options=options)

File: packages/core/src/mcp/drivers.ts
/** Driver factory functions for different platforms. */

import path from "node:path";
import {
  chromium,
  type Browser,
  type BrowserContext,
  type Page,
} from "playwright";
import { Builder, type WebDriver } from "selenium-webdriver";
import { Options } from "selenium-webdriver/chrome.js";
import { getLogger } from "../utils/logger.js";

const logger = getLogger(import.meta.url);

export type McpDriver = Page | WebDriver;

export namespace McpDriver {
  export type PlaywrightCookie = Parameters<
    BrowserContext["addCookies"]
  >[0][number];

  export type PlaywrightHeaders = (Parameters<
    Browser["newContext"]
  >[0] & {})["extraHTTPHeaders"] & {};

  export interface Capabilities {
    headers?: PlaywrightHeaders | undefined;
    cookies?: PlaywrightCookie[] | undefined;
    permissions?: string[] | undefined;
    "appium:settings"?: Record<string, unknown> | undefined;
    [key: string]: unknown;
  }

  export interface SeleniumCdpConnection {
    send(method: string, params: Record<string, unknown>): Promise<unknown>;
  }
}

export function createChromeDriver(
  capabilities: McpDriver.Capabilities,
  serverUrl: string | null | undefined,
  artifactsDir: string,
): Promise<McpDriver> {
  const driverType = (process.env.ALUMNIUM_DRIVER || "selenium").toLowerCase();
  logger.info(`Creating Chrome driver using ${driverType}`);
  if (driverType === "playwright") {
    return createPlaywrightDriver(capabilities, artifactsDir);
  } else {
    return createSeleniumDriver(capabilities, serverUrl);
  }
}

/**
 * Create Playwright driver from capabilities.
 */
export async function createPlaywrightDriver(
  capabilities: McpDriver.Capabilities,
  artifactsDir: string,
): Promise<Page> {
  const headless =
    (process.env.ALUMNIUM_PLAYWRIGHT_HEADLESS || "true").toLowerCase() ===
    "true";
  logger.info(`Creating Playwright driver (headless=${headless})`);

  const browser = await chromium.launch({ headless });

  const headers = capabilities["headers"] || {};
  if (Object.keys(headers).length) {
    logger.debug("Setting extra HTTP headers: {headers}", { headers });
  }

  const context: BrowserContext = await browser.newContext({
    recordVideo: { dir: path.join(artifactsDir, "videos") },
    extraHTTPHeaders: headers,
  });

  await context.tracing.start({
    screenshots: true,
    snapshots: true,
    sources: true,
  });

  const { cookies, permissions } = capabilities;

  if (cookies) {
    logger.debug("Adding cookies: {cookies}", { cookies });
    for (const cookie of cookies) {
      cookie["path"] ??= "/";
    }
    await context.addCookies(cookies);
  }

  if (permissions) {
    logger.debug("Granting permissions: {permissions}", { permissions });
    await context.grantPermissions(permissions);
  }

  const page = await context.newPage();

  logger.debug("Playwright driver created successfully");
  return page;
}

/**
 * Create Selenium Chrome driver from capabilities.
 */
export async function createSeleniumDriver(
  capabilities: McpDriver.Capabilities,
  serverUrl: string | null | undefined,
): Promise<WebDriver> {
  logger.info(`Creating Selenium driver (serverUrl=${serverUrl || "local"})`);

  const { headers, cookies, ...seleniumCapabilities } = capabilities;

  const options = new Options();

  // Apply all capabilities to options
  for (const [key, value] of Object.entries(seleniumCapabilities)) {
    if (key !== "platformName") {
      options.set(key, value);
    }
  }

  // Use remote driver if serverUrl provided, otherwise local Chrome
  const builder = new Builder().forBrowser("chrome").setChromeOptions(options);
  if (serverUrl) {
    builder.usingServer(serverUrl);
  }
  const driver = await builder.build();
  const cdp: McpDriver.SeleniumCdpConnection =
    await driver.createCDPConnection("page");

  const cdpPromises: Promise<unknown>[] = [];

  if (headers || cookies) {
    cdpPromises.push(cdp.send("Network.enable", {}));
  }

  if (headers) {
    logger.debug("Setting extra HTTP headers: {headerNames}", {
      headerNames: Object.keys(headers),
    });
    cdpPromises.push(cdp.send("Network.setExtraHTTPHeaders", { headers }));
  }

  if (cookies) {
    logger.debug(`Adding ${cookies.length} cookie(s)`);
    cdpPromises.push(cdp.send("Network.setCookies", { cookies }));
  }

  await Promise.all(cdpPromises);

  logger.debug("Selenium driver created successfully");
  return driver;
}

export async function createIosDriver(
  capabilities: McpDriver.Capabilities,
  serverUrl: string | null | undefined,
): Promise<unknown> {
  return {};
}

export async function createAndroidDriver(
  capabilities: McpDriver.Capabilities,
  serverUrl: string | null | undefined,
): Promise<unknown> {
  return {};
}
